docs/devel/consumers-claims.md

```diff
@@ -0,0 +1,309 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/devel/consumers-claims.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# The consumer/claim model for cluster resources
+
+A pattern has emerged for managing certain classes of resources (e.g. storage
+and networking) in the Kubernetes API.  This doc aims to document this pattern
+for future reuse.
+
+**Table of Contents**
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [The consumer/claim model for cluster resources](#the-consumerclaim-model-for-cluster-resources)
+  - [Background](#background)
+  - [Status](#status)
+  - [General model](#general-model)
+  - [Storage](#storage)
+    - [PersistentVolumeClaim](#persistentvolumeclaim)
+    - [VolumeManagers](#volumemanagers)
+    - [Config](#config)
+    - [Scrubbing](#scrubbing)
+  - [Networking](#networking)
+    - [IngressClaim](#ingressclaim)
+    - [LoadBalancerManagers](#loadbalancermanagers)
+    - [Config](#config)
+    - [Scrubbing](#scrubbing)
+  - [Future](#future)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Background
+
+Some types of resources which are used in a cluster follow a common lifecycle:
+they are provisioned (whether manually or automatically), used by some
+consumer(s), eventually released by the consumer, scrubbed, and finally either
+released or reused.  These resources generally map to discrete physical or
+virtual _things_.
+
+Examples:
+   * Virtual storage (e.g. cloud-provider-hosted block devices)
+   * Physical storage (e.g. partitions on an iSCSI server or NFS shares)
+   * Virtual networking (e.g. cloud-provider-hosted load-balancers)
+   * Physical networking (e.g. VIPs on hardware load-balancers)
+
+These resources may have long lead times to provision (e.g. file a ticket to
+have an admin do work), may have requirements around de-provisioning (e.g.
+disks must be erased), or cost real money (e.g. for-pay cloud IaaS).  To use
+these things in Kubernetes we need a way to represent the in our API, to claim
+them as in-use, and to dynamically provision them (when supported).
+
```

Comment:
+1. I also commented on this above.

I envision supporting a few flavors of exclusive pods, sole-tenancy, equivalence classes of acceptable pods, etc.